### PR TITLE
upcoming: [DI-20348] - Changes for Adding CloudPulseCustomSelect component and Integration

### PR DESCRIPTION
**Pull request by venkymano**
## Description 📝
Changes for new component CloudPulseCustomSelect that helps to display various filters in the dashboard view and integration of custom select component with metrics API call.

## Changes  🔄
List any change relevant to the reviewer.
1. Created new component CloudPulseCustomSelect.
2. Integration of the select custom filters with the metrics API call.

## Target release date 🗓️
9/3/24

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
|Screenshot 2024-08-21 at 12 55 32 PM|Screenshot 2024-08-21 at 12 55 51 PM|
|Screenshot 2024-08-21 at 12 55 59 PM|Screenshot 2024-08-21 at 12 56 12 PM|
|Screenshot 2024-08-21 at 12 56 23 PM|Screenshot 2024-08-21 at 5 19 25 PM|
Screenshot 2024-08-21 at 1 02 02 PM|
|Screenshot 2024-08-21 at 1 10 03 PM|Screenshot 2024-08-21 at 1 10 35 PM|

## How to test 🧪
1. Login as mock user since some of the API endpoints are not available.
2. Go the Monitors tab.
3. In the dashboard selection dropdown, select Dbaas Dashboard.
4. In the Filters selection section, you can see select an Engine dropdown, that uses CloudPulseCustomSelect component.
5. Also you can add a filter in the FilterConfig.ts file for dbaas like below and upon selection, we can see the filter goes in the metrics API request inside filters object.
```
{
      configuration: {
        filterKey: 'test',
        filterType: 'string',
        isFilterable: true, // isFilterable -- this determines whethere you need to pass it metrics api
        isMetricsFilter: false, // if it is false, it will go as a part of filter params, else global filter
        isMultiSelect: false,
        name: 'Test Filter',
        neededInServicePage: false,
        options: [
          {
            id: 'test',
            label: 'Test1',
          },
          {
            id: 'test2',
            label: 'Test2',
          },
        ],
        placeholder: 'Select a Test',
        priority: 5,
        type: CloudPulseSelectTypes.static,
      },
      name: 'Test Filter',
    }
```
6. In the above config, if you remove the options, the error message will appear below the dropdown
**Note-** You might encounter flickering issues, that is caused due to preferences, that will be solved in the upcoming PR

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    };
+
+    let staticErrorText = '';
+
+    // check for input prop errors
+    if (
+      (CloudPulseSelectTypes.static === type &&
+      (!options || options.length === 0)) || (CloudPulseSelectTypes.dynamic === type && !apiV4QueryKey)
+    ) {
+      staticErrorText = 'Pass pass either options or API query key';
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelect.tsx:182]: Is there a typo in this string?

File: packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelectUtils.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    | CloudPulseServiceTypeFiltersOptions[]
+    | null;
+}
+
+/**
+ * This function returns the default selections based on the user preference and options listed
+ * @param defaultSelectionProps - The props needed for getting the default selections
+ * @returns
+ */
+export const getDefaultSelectionsFromPreferencesAndPublishSelectionChanges = (
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelectUtils.ts:74]: Can we shorten the name of this function? Use the JSDoc comment instead to describe what it does.

File: packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelectUtils.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      : selectedValues[0]
+    : undefined;
+};
+
+/**
+ * This functions calls the selection change callback and updates the latest selected filter in the preferences
+ * @param selectionChangeProps - The props needed for selecting the new filter and updating the global preferences
+ */
+
+export const callSelectionChangeAndUpdateGlobalFilters = (
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/CloudPulse/shared/CloudPulseCustomSelectUtils.ts:131]: Same comment here about shortening this function name.

**Comment by venkymano** [comment]
@hkhalil-akamai , addressed the comments, please check if it is good now

**Comment by jaalah-akamai** [comment]
@dwiley-akamai this is ready for review 👍

File: packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts
Comment (dwiley-akamai) [inline packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:1]: @venkymano do you have ESLint enabled in your IDE? Let's try to minimize the linter warnings and ensure the imports and code formatting match our preferred conventions, across the files this PR touches.

Screenshot 2024-08-23 at 3 56 17 PM

File: packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts [hunk 4]
```diff
@@ -206,3 +253,77 @@ export const checkIfAllMandatoryFiltersAreSelected = (
...
+  return filters;
+};
+
+/**
+ * 
+ * @param filterKey The filterKey of the actual filter
+ * @param dashboard The selected dashboard from the global filter view
+ * @returns The filterKeys that needs to be removed from the preferences
+ */
+const getDependantFiltersByFilterKey =
```
Comment (dwiley-akamai) [inline packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:312]:
```suggestion
const getDependentFiltersByFilterKey =
```

File: packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts
Comment (venkymano) [reply packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:1]: @dwiley-akamai , I have the eslint enabled as well, but it shows no warning for me...is there something i am missing

Screenshot 2024-08-24 at 7 24 38 AM

**Comment by venkymano** [comment]
@dwiley-akamai , Apart from eslint, which i am checking once more, i have addressed all your comments. Please check if it is good now.

File: packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts
Comment (venkymano) [reply packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:1]: Fixed it

**Review by hkhalil-akamai — APPROVED** [review]
Thanks for addressing feedback & apologies for the delay in approving.

**Comment by nikhagra** [comment]
> Thanks for addressing feedback & apologies for the delay in approving.

Thank you for approval @hkhalil-akamai.  @jaalah-akamai / @dwiley-akamai can we get second approval and merge if everything is fine

**Review by dwiley-akamai — APPROVED** [review]

**Comment by venkymano** [comment]
@jaalah-akamai , we have enough approvals, please check if it is good for merging

**Pull request merged**